lib/session.ts

```diff
@@ -0,0 +1,348 @@
+        }
+        if (options.onRejected) {
+          rheaSender.on(SenderEvents.rejected, options.onRejected);
+        }
+        if (options.onReleased) {
+          rheaSender.on(SenderEvents.released, options.onReleased);
+        }
+        if (options.onModified) {
+          rheaSender.on(SenderEvents.modified, options.onModified);
+        }
+      }
+
+      const removeListeners = () => {
+        clearTimeout(waitTimer);
+        rheaSender.removeListener(SenderEvents.senderOpen, onSendable);
+        rheaSender.removeListener(SenderEvents.senderClose, onClose);
+      };
+
+      onSendable = (context: rhea.EventContext) => {
+        removeListeners();
+        process.nextTick(() => {
+          log.session("[%s] Resolving the promise with amqp sender '%s'.",
+            this.connection.id, rheaSender.name);
+          resolve(sender);
+        });
+      };
+
+      onClose = (context: rhea.EventContext) => {
+        removeListeners();
+        log.error("[%s] Error occurred while creating a sender over amqp connection: %O.",
+          this.connection.id, context.sender!.error);
+        reject(context.sender!.error);
+      };
+
+      const actionAfterTimeout = () => {
+        removeListeners();
+        const msg: string = `Unable to create the amqp sender ${rheaSender.name} due to ` +
+          `operation timeout.`;
+        log.error("[%s] %s", this.connection.id, msg);
+        reject(new Error(msg));
+      };
+
+      rheaSender.once(SenderEvents.sendable, onSendable);
+      rheaSender.once(SenderEvents.senderClose, onClose);
+      waitTimer = setTimeout(actionAfterTimeout, defaultOperationTimeoutInSeconds * 1000);
+    });
+  }
+
+  /**
+   * Registers the given handler for the specified event.
+   * @param {rhea.SessionEvents} event The event for which the handler needs to be registered.
+   * @param {rhea.OnAmqpEvent} handler The handler function that needs to be executed when the event
+   * occurs
+   * @returns {void} void.
+   */
+  registerHandler(event: rhea.SessionEvents, handler: rhea.OnAmqpEvent): void {
+    this._session.on(event, handler);
+  }
+
+  /**
+   * Removes the given handler for the specified event.
+   * @param {rhea.SessionEvents} event The event for which the handler needs to be removed.
+   * @param {rhea.OnAmqpEvent} handler The handler function that needs to be removed.
+   * @returns {void} void.
+   */
+  removeHandler(event: rhea.SessionEvents, handler: rhea.OnAmqpEvent): void {
+    this._session.removeListener(event, handler);
+  }
+
+  /**
+   * Registers the given handler for the specified event on the connection object.
+   * @param {rhea.ConnectionEvents} event The event for which the handler needs to be registered.
+   * @param {rhea.OnAmqpEvent} handler The handler function that needs to be executed when the event
+   * occurs
+   * @returns {void} void.
+   */
+  registerConnectionHandler(event: rhea.ConnectionEvents, handler: rhea.OnAmqpEvent): void {
```

Comment:
same here. i wouldn't try to shadow specific APIs on the parent object because it becomes a slippery slope and you end up with dilemmas like whether you should also be able to register connection listeners on links, for instance.